### PR TITLE
Match ergonomics 2024: Implement TC's match ergonomics proposal

### PR DESCRIPTION
**Pull request by Jules-Bertholet**
Under gate `ref_pat_eat_one_layer_2024_structural`. Enabling `ref_pat_eat_one_layer_2024` at the same time allows the union of what the individual gates allow. @traviscross

r? @Nadrieril

cc https://github.com/rust-lang/rust/issues/123076

@rustbot label A-edition-2024 A-patterns

### CODE REVIEW

**Comment by Jules-Bertholet** [comment]
It works on my machine?

**Comment by compiler-errors** [comment]
@Jules-Bertholet: That's a debug assertion. You need to add:

```
rust.debug = true
```

To your `config.toml`, since debug assertions are not enabled by default even for the `"compiler"` profile.

File: compiler/rustc_feature/src/unstable.rs [hunk 1]
```diff
@@ -573,6 +573,8 @@ declare_features! (
     (unstable, raw_ref_op, "1.41.0", Some(64490)),
     /// Makes `&` and `&mut` patterns eat only one layer of references in Rust 2024.
     (incomplete, ref_pat_eat_one_layer_2024, "1.79.0", Some(123076)),
+    /// Makes `&` and `&mut` patterns eat only one layer of references in Rust 2024—structural variant
+    (incomplete, ref_pat_eat_one_layer_2024_structural, "1.79.0", Some(123076)),
```
Comment (Nadrieril) [inline compiler/rustc_feature/src/unstable.rs:577]:
```suggestion
    (incomplete, ref_pat_eat_one_layer_2024_structural, "CURRENT_RUSTC_VERSION", Some(123076)),
```

**Comment by Nadrieril** [comment]
Looks good! Just one procedural nit and then you can merge

@bors delegate

**Comment by Nadrieril** [comment]
Forgot I can push that kind of change myself

**Comment by Jules-Bertholet** [comment]
Failure seems unrelated to PR
@rustbot label -S-waiting-on-author

File: tests/ui/match/ref_pat_eat_one_layer_2024/ref_pat_eat_one_layer_2024_fail.rs [hunk 2]
```diff
@@ -50,6 +52,10 @@ pub fn main() {
     let &mut &mut &mut &mut _ = &mut &&&&mut &&&mut &mut 0;
     //~^ ERROR: mismatched types
 
+    if let Some(&mut _) = &mut Some(&0) {
+        //[structural]~^ ERROR
+    }
```
Comment (traviscross) [inline tests/ui/match/ref_pat_eat_one_layer_2024/ref_pat_eat_one_layer_2024_fail.rs:57]: This one will start working when we change to the revised (extended) *Rule 4*.

**Review by traviscross — APPROVED** [review]
Looking at the tests that changed here, this looks right to me.

File: tests/ui/match/ref_pat_eat_one_layer_2024/ref_pat_eat_one_layer_2024_fail.rs (hunk 2 above)
Comment (Jules-Bertholet) [reply tests/ui/match/ref_pat_eat_one_layer_2024/ref_pat_eat_one_layer_2024_fail.rs:57]: Yes, hence why it is marked as failing in the `structural` variant only (not `both` which is the revised rule).

Comment (traviscross) [reply tests/ui/match/ref_pat_eat_one_layer_2024/ref_pat_eat_one_layer_2024_fail.rs:57]: Makes sense.

**Pull request merged**